File: Python/space/basis.py
from functools import lru_cache

import numpy as np
import quadpy

from ..datastructures.function import FunctionInterface
from ..datastructures.tree import MetaRoot
from ..datastructures.tree_view import NodeView, TreeView
from .triangulation import Vertex


@lru_cache(maxsize=10)
def _get_quadrature_scheme(order):
    # order == 2 * s + 1.
    return quadpy.nsimplex.grundmann_moeller(2, order // 2)


class HierarchicalBasisFunction(FunctionInterface, NodeView):
    order = 1

    def __init__(self, nodes, parents=None, children=None):
        super().__init__(nodes=nodes, parents=parents, children=children)
        assert isinstance(self.node, (Vertex, MetaRoot))

    @property
    def on_domain_boundary(self):
        return self.node.on_domain_boundary

    @property
    def support(self):
        return self.node.patch

    @property
    def patch(self):
        return self.node.patch

    def eval(self, xy, deriv=False):
        """ Evaluate hat function on a number of points `xy` at once. """
        assert xy.shape[0] == 2
        # If we input a single vector xy, we expect a number out.
        if len(xy.shape) == 1:
            result = np.zeros(xy.shape) if deriv else np.zeros(1)
        else:
            result = np.zeros(xy.shape) if deriv else np.zeros(xy.shape[1])
        for elem in self.support:
            i = elem.vertices.index(self.node)
            bary = elem.to_barycentric_coordinates(xy)
            # mask[j] == True exactly when point xy[:,j] is inside elem.
            mask = np.all(bary >= 0, axis=0)
            if not any(mask):
                continue
            if not deriv:
                result[mask] = bary[i, mask]
            else:
                V = elem.vertex_array().T
                opp_edge = V[:, (i - 1) % 3] - V[:, (i + 1) % 3]
                normal = np.array([-opp_edge[1], opp_edge[0]])
                normal = -normal / (2 * elem.area)
                result[:, mask] = normal.reshape(2, 1)
        # Return singular float if the input was a singular vector xy.
        return result if len(xy.shape) == 2 else result[0]

    def inner_quad(self, g, g_order=2, deriv=False):
        """ Computes <g, self> or <g, grad self> by quadrature. """
        if not deriv:
            func = lambda xy: self.eval(xy, deriv) * g(xy)
        else:  # g is a vector field, so take the inner product.
            func = lambda xy: (self.eval(xy, deriv) * g(xy)).sum(axis=0)
        scheme = _get_quadrature_scheme(g_order + self.order)
        result = 0.0
        for elem in self.support:
            triangle = elem.vertex_array()
            result += scheme.integrate(func, triangle)
        return result

    @staticmethod
    def from_triangulation(triangulation):
        """ Creates hierarchical basis function tree from the given triang. """
        return TreeView(
            HierarchicalBasisFunction(triangulation.vertex_meta_root))

File: src/space/basis.cpp
#include "basis.hpp"

namespace space {

bool HierarchicalBasisFn::Refine() {
  if (is_full()) return false;
  for (auto &elem : vertex_->patch) {
    elem->Refine();
    assert(elem->children().size() == 2);
    elem->children()[0]->newest_vertex()->RefineHierarchicalBasisFn();
  }
  assert(is_full());
  return true;
}

bool HierarchicalBasisFn::is_full() const {
  if (is_metaroot()) return vertex_->children().size() == children().size();
  for (auto &elem : vertex_->patch)
    if (!elem->is_full()) return false;
  return true;
}

double HierarchicalBasisFn::Volume() const {
  double vol = 0.0;
  for (auto elem : support()) vol += elem->area();
  return vol;
}

double HierarchicalBasisFn::Eval(double x, double y) const {
  for (auto elem : support()) {
    auto bary = elem->BarycentricCoordinates(x, y);

    // Check if the point is contained inside this element.
    if ((bary.array() >= 0).all()) {
      // Find which barycentric coordinate corresponds to this hat fn.
      for (int i = 0; i < 3; ++i)
        if (elem->vertices()[i] == vertex()) return bary[i];
      assert(false);
    }
  }
  return 0;
}

bool HierarchicalBasisFn::Contains(double x, double y) const {
  for (auto elem : support()) {
    auto bary = elem->BarycentricCoordinates(x, y);
    if ((bary.array() >= 0).all()) return true;
  }
  return false;
}

Eigen::Vector2d HierarchicalBasisFn::EvalGrad(double x, double y) const {
  for (auto elem : support()) {
    auto bary = elem->BarycentricCoordinates(x, y);

    // Check if the point is contained inside this element.
    if ((bary.array() >= 0).all()) {
      // Find which barycentric coordinate corresponds to this hat fn.
      for (int i = 0; i < 3; ++i)
        if (elem->vertices()[i] == vertex()) {
          auto [vp1, vm1] = elem->edge(i);
          Eigen::Vector2d normal{vm1->y - vp1->y, vp1->x - vm1->x};
          normal /= 2.0 * elem->area();
          return normal;
        }
      assert(false);
    }
  }
  return {0, 0};
}

}  // namespace space
